File: examples/checksummed_cache.c
/*
 * Checksummed block with per-thread in-memory cache protected by CRDS.
 *
 * The bstack_t pointer is shared across threads.  The per-thread cache
 * (cached_checksum, cached_payload) lives in each thread's own
 * checksummed_block_t, constructed inside the thread function.
 *
 * A pthread_barrier forces thread 2 to attempt its write only after thread 1
 * has already committed, guaranteeing that thread 2's empty cache is stale.
 * 
 * Note that XOR checksums are not collision-resistant, so this example is only
 * intended to demonstrate the pattern, not to provide strong integrity guarantees.
 *
 * Requires: -DBSTACK_FEATURE_SET -DBSTACK_FEATURE_ATOMIC
 *
 * Build and run:
 *   make -C ../c example-checksummed_cache
 */

#if !defined(BSTACK_FEATURE_SET) || !defined(BSTACK_FEATURE_ATOMIC)
#  error "checksummed_cache.c requires -DBSTACK_FEATURE_SET -DBSTACK_FEATURE_ATOMIC"
#endif

#include "../c/bstack.h"

#include <pthread.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define BLOCK_SIZE     64
#define PAYLOAD_OFFSET 8
#define PAYLOAD_SIZE   56

/* ── Little-endian helper ────────────────────────────────────────────────── */

static void write_le64(uint8_t *p, uint64_t v)
{
    p[0] = (uint8_t)(v);       p[1] = (uint8_t)(v >>  8);
    p[2] = (uint8_t)(v >> 16); p[3] = (uint8_t)(v >> 24);
    p[4] = (uint8_t)(v >> 32); p[5] = (uint8_t)(v >> 40);
    p[6] = (uint8_t)(v >> 48); p[7] = (uint8_t)(v >> 56);
}

/* ── ChecksummedBlock ────────────────────────────────────────────────────── */

typedef struct {
    bstack_t *stack;                    /* shared across threads */
    int       has_checksum;
    uint8_t   cached_block[BLOCK_SIZE]; /* [0..8) checksum | [8..64) payload */
} checksummed_block_t;

static uint64_t compute_checksum(const uint8_t *payload, size_t len)
{
    uint64_t cs = 0;
    for (size_t i = 0; i < len; i++)
        cs ^= (uint64_t)payload[i];
    return cs;
}

/*
 * Write new_payload atomically via eq_crds.
 *
 * *committed is set to 1 if the write was committed, or 0 if the cached
 * checksum was stale.  On a stale miss the cache is refreshed via bstack_get
 * so the caller can retry immediately.
 */
static int cb_write(checksummed_block_t *cb,
                    const uint8_t *new_payload, size_t len,
                    int *committed)
{
    uint64_t new_checksum = compute_checksum(new_payload, len);

    uint8_t block[BLOCK_SIZE] = {0};
    write_le64(block, new_checksum);
    memcpy(block + PAYLOAD_OFFSET, new_payload, len);

    /* Use cached_block[0..8] as expected checksum (zeroes if no cache yet).
     * cached_block also serves as b_old_buf: on match, eq_crds fills it with
     * the old on-disk block before writing the new one. */
    static const uint8_t zeroes[8] = {0};
    const uint8_t *expected = cb->has_checksum ? cb->cached_block : zeroes;

    int ok = 0;
    if (bstack_eq_crds(cb->stack,
                       0, expected, 8,
                       0, cb->cached_block, block, BLOCK_SIZE,
                       &ok) < 0)
        return -1;

    if (ok) {
        /* Committed — overwrite cached_block with the new state. */
        memcpy(cb->cached_block, block, BLOCK_SIZE);
        cb->has_checksum = 1;
        *committed = 1;
        return 0;
    }

    /* Stale — read current disk state directly into cached_block; no copy. */
    if (bstack_get(cb->stack, 0, BLOCK_SIZE, cb->cached_block) < 0)
        return -1;
    cb->has_checksum = 1;
    *committed = 0;
    return 0;
}

/* ── Read generator ──────────────────────────────────────────────────────── */

typedef struct {
    checksummed_block_t *cb;
    int     need_checksum;
    uint8_t read_checksum[8];
} read_ctx_t;

static int read_gen(uint64_t *out_offset, uint8_t **out_buf, size_t *out_len,
                    void *userctx)
{
    read_ctx_t          *ctx = userctx;
    checksummed_block_t *cb  = ctx->cb;

    if (ctx->need_checksum) {
        ctx->need_checksum = 0;
        *out_offset = 0;
        *out_buf    = ctx->read_checksum;
        *out_len    = 8;
        return 1;
    }

    /* Cache hit — checksum unchanged, skip payload fetch. */
    if (cb->has_checksum &&
        memcmp(ctx->read_checksum, cb->cached_block, 8) == 0)
        return 0;

    /* Cache miss — update checksum bytes and fetch payload directly into
     * cached_block; no copy needed for the payload. */
    memcpy(cb->cached_block, ctx->read_checksum, 8);
    cb->has_checksum = 1;
    *out_offset = PAYLOAD_OFFSET;
    *out_buf    = cb->cached_block + PAYLOAD_OFFSET;
    *out_len    = PAYLOAD_SIZE;
    return 1;
}

static int cb_read(checksummed_block_t *cb, const uint8_t **out_payload)
{
    read_ctx_t ctx = { .cb = cb, .need_checksum = 1, .read_checksum = {0} };
    if (bstack_get_batched_gen(cb->stack, read_gen, &ctx) < 0)
        return -1;
    *out_payload = cb->cached_block + PAYLOAD_OFFSET;
    return 0;
}

/* ── Helpers ─────────────────────────────────────────────────────────────── */

static void print_trimmed(const uint8_t *payload, size_t len)
{
    size_t trimmed = 0;
    for (size_t i = 0; i < len; i++)
        if (payload[i] != 0) trimmed = i + 1;
    printf("\"%.*s\"", (int)trimmed, (const char *)payload);
}

/* ── Thread functions ────────────────────────────────────────────────────── */

typedef struct {
    bstack_t       *stack;
    pthread_mutex_t mtx;
    pthread_cond_t  cond;
    int             t1_written; /* set to 1 after thread 1 commits its write */
} thread_arg_t;

static void *thread1(void *arg)
{
    thread_arg_t        *a  = arg;
    checksummed_block_t  cb = { .stack = a->stack, .has_checksum = 0, .cached_block = {0} };

    int committed = 0;
    if (cb_write(&cb, (const uint8_t *)"Written by thread 1.", 20, &committed) < 0) {
        perror("[t1] write"); return NULL;
    }
    printf("[t1] write          → committed=%d\n", committed);

    const uint8_t *payload = NULL;
    if (cb_read(&cb, &payload) < 0) { perror("[t1] read"); return NULL; }
    printf("[t1] read           → hit  "); print_trimmed(payload, PAYLOAD_SIZE); putchar('\n');

    /* Signal thread 2 that the write has landed. */
    pthread_mutex_lock(&a->mtx);
    a->t1_written = 1;
    pthread_cond_signal(&a->cond);
    pthread_mutex_unlock(&a->mtx);
    return NULL;
}

static void *thread2(void *arg)
{
    thread_arg_t        *a  = arg;
    checksummed_block_t  cb = { .stack = a->stack, .has_checksum = 0, .cached_block = {0} };

    /* Wait until thread 1 has committed its write; our cache (None) is stale. */
    pthread_mutex_lock(&a->mtx);
    while (!a->t1_written)
        pthread_cond_wait(&a->cond, &a->mtx);
    pthread_mutex_unlock(&a->mtx);

    int committed = 0;
    if (cb_write(&cb, (const uint8_t *)"Written by thread 2.", 20, &committed) < 0) {
        perror("[t2] write (stale)"); return NULL;
    }
    printf("[t2] write (stale)  → committed=%d, cache refreshed\n", committed);

    /* Retry — cache is now current. */
    if (cb_write(&cb, (const uint8_t *)"Written by thread 2.", 20, &committed) < 0) {
        perror("[t2] write (retry)"); return NULL;
    }
    printf("[t2] write (retry)  → committed=%d\n", committed);

    const uint8_t *payload = NULL;
    if (cb_read(&cb, &payload) < 0) { perror("[t2] read"); return NULL; }
    printf("[t2] read           → hit  "); print_trimmed(payload, PAYLOAD_SIZE); putchar('\n');

    return NULL;
}

/* ── Main ────────────────────────────────────────────────────────────────── */

int main(void)
{
    const char *path = "checksummed_cache_example.bstack";
    remove(path);

    bstack_t *stack = bstack_open(path);
    if (!stack) { perror("bstack_open"); return 1; }

    /* Initialise with a zeroed block. */
    uint8_t zeros[BLOCK_SIZE] = {0};
    if (bstack_push(stack, zeros, BLOCK_SIZE, NULL) < 0) {
        perror("bstack_push"); bstack_close(stack); return 1;
    }

    thread_arg_t arg = {
        .stack      = stack,
        .mtx        = PTHREAD_MUTEX_INITIALIZER,
        .cond       = PTHREAD_COND_INITIALIZER,
        .t1_written = 0,
    };

    pthread_t t1, t2;
    pthread_create(&t1, NULL, thread1, &arg);
    pthread_create(&t2, NULL, thread2, &arg);
    pthread_join(t1, NULL);
    pthread_join(t2, NULL);

    pthread_mutex_destroy(&arg.mtx);
    pthread_cond_destroy(&arg.cond);
    bstack_close(stack);
    return 0;
}

File: examples/checksummed_cache.rs
//! Checksummed block with in-memory cache protected by CRDS operations.
//!
//! ## Layout
//!
//! A single 64-byte block stored at offset 0:
//! - Bytes [0..8): 8-byte XOR checksum (LE u64)
//! - Bytes [8..64): 56-byte payload
//!
//! ## Write path
//!
//! Use `eq_crds` to compare the cached checksum against the on-disk checksum
//! and, if they match, atomically replace the entire block.  On a mismatch the
//! stale cache is refreshed via `get_into` and the call returns `false` so the
//! caller can retry.
//!
//! ## Read path
//!
//! Use `get_batched_gen` to read the checksum and, only on a mismatch with the
//! cached value, fetch the new payload — both under a single read lock.
//! 
//! Note that XOR checksums are not collision-resistant, so this example is only
//! intended to demonstrate the pattern, not to provide strong integrity guarantees.
//!
//! ## Thread model
//!
//! The underlying `BStack` is wrapped in an `Arc` and shared across threads.
//! The per-thread cache (`cached_checksum`, `cached_payload`) lives in each
//! thread's own `ChecksummedBlock` instance, constructed inside the thread.
//!
//! ## How to run
//!
//! ```text
//! cargo run --example checksummed_cache --features "set,atomic"
//! ```

#[cfg(all(feature = "set", feature = "atomic"))]
use bstack::BStack;
#[cfg(all(feature = "set", feature = "atomic"))]
use std::io;
#[cfg(all(feature = "set", feature = "atomic"))]
use std::sync::{Arc, Barrier};
#[cfg(all(feature = "set", feature = "atomic"))]
use std::thread;

#[cfg(all(feature = "set", feature = "atomic"))]
const BLOCK_SIZE: usize = 64;
#[cfg(all(feature = "set", feature = "atomic"))]
const PAYLOAD_OFFSET: u64 = 8;
#[cfg(all(feature = "set", feature = "atomic"))]
const PAYLOAD_SIZE: usize = 56;

#[cfg(all(feature = "set", feature = "atomic"))]
struct ChecksummedBlock {
    stack: Arc<BStack>,
    /// Per-thread cache — each thread owns its own ChecksummedBlock.
    cached_checksum: Option<[u8; 8]>,
    cached_payload: [u8; PAYLOAD_SIZE],
}

#[cfg(all(feature = "set", feature = "atomic"))]
impl ChecksummedBlock {
    /// Initialise the backing file and return a shared stack handle.
    fn open(path: &str) -> io::Result<Arc<BStack>> {
        let _ = std::fs::remove_file(path);
        let stack = BStack::open(path)?;
        stack.push(&[0u8; BLOCK_SIZE])?;
        Ok(Arc::new(stack))
    }

    /// Create a per-thread view over a shared stack.
    fn new(stack: Arc<BStack>) -> Self {
        Self {
            stack,
            cached_checksum: None,
            cached_payload: [0u8; PAYLOAD_SIZE],
        }
    }

    /// Compute XOR checksum of the payload.
    fn compute_checksum(payload: &[u8]) -> u64 {
        payload.iter().fold(0u64, |acc, &byte| acc ^ (byte as u64))
    }

    /// Write `new_payload` atomically.
    ///
    /// Returns `true` if the write was committed, or `false` if the cached
    /// checksum was stale.  On a stale miss the cache is refreshed from the
    /// file via `get_into` so the caller can retry immediately.
    fn write(&mut self, new_payload: &[u8]) -> io::Result<bool> {
        assert!(new_payload.len() <= PAYLOAD_SIZE);

        let new_checksum = Self::compute_checksum(new_payload);
        let mut block = [0u8; BLOCK_SIZE];
        block[0..8].copy_from_slice(&new_checksum.to_le_bytes());
        block[PAYLOAD_OFFSET as usize..PAYLOAD_OFFSET as usize + new_payload.len()]
            .copy_from_slice(new_payload);

        // Verify cached checksum (or expect zeroed checksum for first write).
        let expected_checksum = self.cached_checksum.unwrap_or([0u8; 8]);

        if self
            .stack
            .eq_crds(0, &expected_checksum, 0, block)?
            .is_some()
        {
            // Committed — update the cache to reflect the new file state.
            self.cached_checksum = Some(new_checksum.to_le_bytes());
            self.cached_payload[..new_payload.len()].copy_from_slice(new_payload);
            self.cached_payload[new_payload.len()..].fill(0);
            return Ok(true);
        }

        // Cache is stale — refresh it so the caller can retry.
        let mut disk = [0u8; BLOCK_SIZE];
        self.stack.get_into(0, &mut disk)?;
        self.cached_checksum = Some(disk[0..8].try_into().unwrap());
        self.cached_payload
            .copy_from_slice(&disk[PAYLOAD_OFFSET as usize..]);
        Ok(false)
    }

    fn read(&mut self) -> io::Result<&[u8]> {
        let mut need_checksum_read = true;
        let mut read_checksum = [0u8; 8];
        self.stack.get_batched_gen(|| {
            if need_checksum_read {
                need_checksum_read = false;
                Some((
                    0,
                    // SAFETY: `read_checksum` lives for the duration of get_batched_gen
                    unsafe { std::slice::from_raw_parts_mut(read_checksum.as_mut_ptr(), 8) },
                ))
            } else {
                if let Some(old_checksum) = self.cached_checksum
                    && read_checksum == old_checksum
                {
                    return None;
                }
                self.cached_checksum = Some(read_checksum);
                Some((
                    PAYLOAD_OFFSET,
                    // SAFETY: `cached_payload` lives for the duration of get_batched_gen
                    unsafe {
                        std::slice::from_raw_parts_mut(
                            self.cached_payload.as_mut_ptr(),
                            PAYLOAD_SIZE,
                        )
                    },
                ))
            }
        })?;
        Ok(&self.cached_payload)
    }
}

/// Return the payload bytes up to the last non-zero byte.
#[cfg(all(feature = "set", feature = "atomic"))]
fn trim(payload: &[u8]) -> &str {
    let len = payload.iter().rposition(|&b| b != 0).map_or(0, |i| i + 1);
    std::str::from_utf8(&payload[..len]).unwrap_or("")
}

#[cfg(all(feature = "set", feature = "atomic"))]
fn main() -> io::Result<()> {
    let stack = ChecksummedBlock::open("checksummed_cache_example.bstack")?;

    // A barrier lets thread 1 commit its write before thread 2 attempts its own,
    // guaranteeing that thread 2's cache is stale when it first tries to write.
    let barrier = Arc::new(Barrier::new(2));

    // ── Thread 1: write first, then release thread 2 ──────────────────────────
    let stack1 = Arc::clone(&stack);
    let barrier1 = Arc::clone(&barrier);
    let t1 = thread::spawn(move || -> io::Result<()> {
        let mut block = ChecksummedBlock::new(stack1);

        assert!(block.write(b"Written by thread 1.")?);
        println!("[t1] write          → committed");

        let payload = block.read()?;
        println!("[t1] read           → hit  {:?}", trim(payload));

        barrier1.wait(); // release thread 2
        Ok(())
    });

    // ── Thread 2: wait for thread 1, then write with a stale cache ────────────
    let stack2 = Arc::clone(&stack);
    let barrier2 = Arc::clone(&barrier);
    let t2 = thread::spawn(move || -> io::Result<()> {
        let mut block = ChecksummedBlock::new(stack2);

        barrier2.wait(); // thread 1 has written by now; our cache (None) is stale

        // First attempt: stale — eq_crds sees a checksum mismatch, cache is
        // refreshed via get_into, returns false.
        let committed = block.write(b"Written by thread 2.")?;
        println!("[t2] write (stale)  → committed={committed}, cache refreshed");

        // Retry: cache is now current, write succeeds.
        let committed = block.write(b"Written by thread 2.")?;
        println!("[t2] write (retry)  → committed={committed}");

        let payload = block.read()?;
        println!("[t2] read           → hit  {:?}", trim(payload));

        Ok(())
    });

    t1.join().unwrap()?;
    t2.join().unwrap()?;
    Ok(())
}

#[cfg(not(all(feature = "set", feature = "atomic")))]
fn main() {
    eprintln!("This example requires the 'set' and 'atomic' features.");
    eprintln!("Run: cargo run --example checksummed_cache --features \"set,atomic\"");
}
